### PR TITLE
API for target properties

### PR DESCRIPTION
**Pull request by lhstrh**
These changes make it easier to add new target properties and type check them. This also adds a suite of unit tests to address #270.

### CODE REVIEW

File: xtext/org.icyphy.linguafranca/src/org/icyphy/generator/CGenerator.xtend [hunk 1]
```diff
@@ -406,8 +406,50 @@ class CGenerator extends GeneratorBase {
                 initializeTriggerObjectsEnd = new StringBuilder()                
                         
                 // Enable clock synchronization if the federate is not local and clock-sync is enabled
+//<<<<<<< HEAD
```
Comment (Soroosh129) [inline xtext/org.icyphy.linguafranca/src/org/icyphy/generator/CGenerator.xtend:409]: This looks like a weird merge conflict. I've refactored everything to `initializeClockSynchronization()` on `master` so I think these can be safely removed.

Comment (lhstrh) [reply xtext/org.icyphy.linguafranca/src/org/icyphy/generator/CGenerator.xtend:409]: Jep

File: xtext/org.icyphy.linguafranca/src/org/icyphy/Target.java [hunk 2]
```diff
@@ -347,104 +361,236 @@
...
         /**
          * Flags to be passed on to the target compiler.
          */
-        FLAGS("flags", Arrays.asList(Target.C, Target.CCpp)),
+        FLAGS("flags", PrimitiveType.STRING,
+                Arrays.asList(Target.C, Target.CCPP), (config, value) -> {
+                    config.compilerFlags.clear();
+                    String str = ASTUtils.toText(value);
+                    if (!str.isEmpty()) {
+                        // FIXME: this is unsafe because it doesn't account for
```
Comment (lhstrh) [inline xtext/org.icyphy.linguafranca/src/org/icyphy/Target.java:503]: @Soroosh129: this came from a merge with code that you recently added. Could you take a stab at addressing this FIXME?

Comment (Soroosh129) [reply xtext/org.icyphy.linguafranca/src/org/icyphy/Target.java:527]: Sure! Should we make it an actual list like the files property?

File: xtext/org.icyphy.linguafranca.tests/src/org/icyphy/tests/LinguaFrancaValidationTest.xtend [hunk 3]
```diff
@@ -787,14 +800,186 @@ class LinguaFrancaValidationTest {
...
+                val examples = newLinkedList
+                if (correct) {
+                    type.options.forEach[
+                    if (it instanceof TargetPropertyType) {
+                        synthesizeExamples(it, correct).forEach[examples.add(it)]
+                    } else {
+                        // It must be a plain Enum<?>
+                        examples.add(it.toString())
+                    }
+                ]
```
Comment (lsk567) [inline xtext/org.icyphy.linguafranca.tests/src/org/icyphy/tests/LinguaFrancaValidationTest.xtend:869]: indentation?

File: xtext/org.icyphy.linguafranca.tests/src/org/icyphy/tests/LinguaFrancaValidationTest.xtend [hunk 4]
```diff
@@ -787,14 +800,186 @@ class LinguaFrancaValidationTest {
...
+                } else {
+                    // Return some obviously incorrect examples for the common
+                    // case where the options are from an ordinary Enum<?>.
+                    if (!type.options.exists[it instanceof TargetPropertyType]) {
+                        return #["foo", "\"bar\"", "'baz'", "1", "-1", "{x: 42}", "[1, 2, 3]"]
```
Comment (lsk567) [inline xtext/org.icyphy.linguafranca.tests/src/org/icyphy/tests/LinguaFrancaValidationTest.xtend:874]: So we are not synthesizing bad examples here? I see [line 864](https://github.com/icyphy/lingua-franca/pull/272/files#diff-523829f2c3f67a249fddb21468b49d2838ea7e3df75e6ac34f56448aca8f3c05R864) calls `synthesizeExample`.

Comment (lsk567) [reply xtext/org.icyphy.linguafranca.tests/src/org/icyphy/tests/LinguaFrancaValidationTest.xtend:874]: Maybe I don't understand the function yet.. What is the intuition behind the recursion?

File: xtext/org.icyphy.linguafranca/src/org/icyphy/Configuration.xtend [hunk 5]
```diff
@@ -35,8 +41,96 @@ import org.icyphy.Target.CoordinationType
...
+    static def boolean fileExists(String filename, String directory) {
+        // Make sure the file exists and issue a warning if not.
+        val file = filename.findFile(directory)
+        if (file === null) {
+            // See if it can be found as a resource.
+            val stream = Configuration.getResourceAsStream(filename)
+            if (stream === null) {
+                // Warn that file hasn't been found.
+                return false
+//                this.reportWarning(value,
```
Comment (lsk567) [inline xtext/org.icyphy.linguafranca/src/org/icyphy/Configuration.xtend:54]: Looks like dead code. Can we remove?

File: xtext/org.icyphy.linguafranca/src/org/icyphy/Target.java (hunk 2 above)
Comment (lhstrh) [reply xtext/org.icyphy.linguafranca/src/org/icyphy/Target.java:527]: Ah, good idea. That would be much better. In that case, I'll take care of it.

File: xtext/org.icyphy.linguafranca/src/org/icyphy/Configuration.xtend [hunk 6]
```diff
@@ -251,5 +249,19 @@ class Configuration {
...
+class ClockSyncOptions {
+    public int attenuation = 10
+    
+    public TimeValue period = new TimeValue(5, TimeUnit.MSEC)
+    
+    public int trials = 10
+    
+    public boolean localFederatesOn
+    
+    public boolean collectStats = true // Should this be the default???
```
Comment (lhstrh) [inline xtext/org.icyphy.linguafranca/src/org/icyphy/Configuration.xtend:264]: @Soroosh129: thanks for spotting this comment and clarifying!

File: xtext/org.icyphy.linguafranca.tests/src/org/icyphy/tests/LinguaFrancaValidationTest.xtend (hunk 4 above)
Comment (lhstrh) [reply xtext/org.icyphy.linguafranca.tests/src/org/icyphy/tests/LinguaFrancaValidationTest.xtend:874]: I'll clean this up a bit and write some more comments. This stuff is a bit idiosyncratic, but it's main purpose is to rely as little as possible on manually provided test vectors. The recursive calls are to deal with composite types (arrays, dictionaries).

File: xtext/org.icyphy.linguafranca/src/org/icyphy/Configuration.xtend (hunk 5 above)
Comment (lhstrh) [reply xtext/org.icyphy.linguafranca/src/org/icyphy/Configuration.xtend:54]: Jep

File: xtext/org.icyphy.linguafranca.tests/src/org/icyphy/tests/LinguaFrancaValidationTest.xtend [hunk 7]
```diff
@@ -787,14 +799,212 @@ class LinguaFrancaValidationTest {
...
     }
     
-    @Test
-    def void checkTargetProperties() {
-        parseWithoutError('''
-                target C {coordination: centralized};
+    /**
+     * Maps a type to a list of known good values.
+     */
+    val primitiveTypeToKnownGood = #{
```
Comment (lhstrh) [inline xtext/org.icyphy.linguafranca.tests/src/org/icyphy/tests/LinguaFrancaValidationTest.xtend:805]: @Soroosh129, @lsk567: these lists deserve scrutiny. I'm pretty certain that these don't cover all the cases. If you find any missing, please add them.

**Pull request merged**